### PR TITLE
Fix completion record for variable declarations

### PR DESCRIPTION
**Pull request by addaleax**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | –
| Patch: Bug Fix?          | Yes
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | –
| Any Dependency Changes?  | –
| License                  | MIT

Variable declarations are ignored for completion records,
so they should be skipped.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/13596"></a>

### CODE REVIEW

**Comment by JLHwung** [comment]
I think we should iterate statement list from end to start, if the completion record is a variable declaration, discard it and check the previous one:

https://github.com/babel/babel/blob/844baebd26547f032b1bb7d38738062ca6f85524/packages/babel-traverse/src/path/family.ts#L195-L201

**Comment by addaleax** [comment]
@JLHwung Sure, done!

File: packages/babel-traverse/src/path/family.ts [hunk 1]
```diff
@@ -193,11 +193,16 @@ function getStatementListCompletion(
...
-    // the completion
-    completions = completions.concat(
-      _getCompletionRecords(paths[paths.length - 1], context),
-    );
+    // search on statement lists and assume that the last
+    // non-variable-declaration statement determines the completion.
+    for (let i = paths.length - 1; i >= 0; i--) {
+      if (!paths[i].isVariableDeclaration()) {
+        completions = completions.concat(
+          _getCompletionRecords(paths[paths.length - 1], context),
```
Comment (JLHwung) [inline packages/babel-traverse/src/path/family.ts:201]:
```suggestion
          _getCompletionRecords(paths[i], context),
```
Guess it should have been `paths[i]`.

File: packages/babel-traverse/src/path/family.ts [hunk 2]
```diff
@@ -193,11 +193,16 @@ function getStatementListCompletion(
...
     // When we are in a context where `break` must not exist, we can skip linear
-    // search on statement lists and assume that the last statement determines
-    // the completion
-    completions = completions.concat(
-      _getCompletionRecords(paths[paths.length - 1], context),
-    );
+    // search on statement lists and assume that the last
+    // non-variable-declaration statement determines the completion.
+    for (let i = paths.length - 1; i >= 0; i--) {
+      if (!paths[i].isVariableDeclaration()) {
```
Comment (JLHwung) [inline packages/babel-traverse/src/path/family.ts:199]: I think we should apply `_getCompletionRecords` and check if the completion is `VariableDeclaration`, e.g. The completion record of this snippet

```js
1 + 2;
{
  var a = 1;
}
```
is `1 + 2`.

Comment (addaleax) [reply packages/babel-traverse/src/path/family.ts]: @JLHwung Sure, done, and added a test. I have to admit that I’m a bit unsure about the code here – is it valid to assume that if the list of completion records returned by `_getCompletionRecords` contains a `VariableDeclaration`, then that automatically implies that the list has length 1? It seems like something that should be true to me, and I’m not sure how to handle this if it isn’t.

Comment (JLHwung) [reply packages/babel-traverse/src/path/family.ts]: > is it valid to assume that if the list of completion records returned by _getCompletionRecords contains a VariableDeclaration, then that automatically implies that the list has length 1

That's a very good question! AFAIK only IfStatement, SwitchStatement and TryStatement may produce multiple completion records, and if a VariableDeclaration is directly within these statements, it should be recursively handled here and as long as we don't return VariableDeclaration as completion records in `getStatementListCompletion`, then the `VariableDeclaration` must _not_ come from these statements and thus the list here should contain only one completion record.

Comment (addaleax) [reply packages/babel-traverse/src/path/family.ts]: Right – that’s why I’d be thinking. :) Should I add an assertion of some kind?

Comment (JLHwung) [reply packages/babel-traverse/src/path/family.ts]: > an assertion of some kind

I think it's fine. The `_getCompletionRecords` is mostly used in `proposal-do-expression`, and the proposal now [bans](https://tc39.es/proposal-do-expressions/#sec-doexpression-static-semantics-early-errors) variable declaration ending in do block. So I guess few people will use it in that way.

**Review by JLHwung — APPROVED** [review]

File: packages/babel-traverse/src/path/family.ts [hunk 3]
```diff
@@ -193,11 +193,19 @@ function getStatementListCompletion(
...
+    // search on statement lists and assume that the last
+    // non-variable-declaration statement determines the completion.
+    for (let i = paths.length - 1; i >= 0; i--) {
+      const pathCompletions = _getCompletionRecords(paths[i], context);
+      if (
+        pathCompletions.length > 1 ||
+        (pathCompletions.length === 1 &&
+          !pathCompletions[0].path.isVariableDeclaration())
+      ) {
+        completions = completions.concat(pathCompletions);
```
Comment (fedeci) [inline packages/babel-traverse/src/path/family.ts:205]: We can improve performances avoiding `.concat` like this:
```suggestion
        completions.push(...pathCompletions);
```

Comment (nicolo-ribaudo) [reply packages/babel-traverse/src/path/family.ts:205]: In this function we always use `.concat`; it should probably be addressed in a refactoring PR where we change all of them.

Comment (fedeci) [reply packages/babel-traverse/src/path/family.ts:205]: I'll prepare it.

**Review by nicolo-ribaudo — APPROVED** [review]
Thanks!

**Pull request merged**